### PR TITLE
Add multi-class classification target

### PR DESCRIPTION
**Pull request by giovannt0**

### CODE REVIEW

File: estimator.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def __init__(self,
                nb_trees_per_ensemble: int,
                max_depth: int,
                learning_rate: float,
+               n_classes: int = None,
```
Comment (giovannt0) [inline estimator.py]: Can you type the argument as `Optional[int]` for consistency?

File: estimator.py [hunk 2]
```diff
@@ -109,16 +112,20 @@ def predict(self, X: np.array) -> np.array:
...
-      return reg_preds
-    class_preds = []
-    for reg in reg_preds:
-      if reg < 0:
-        class_preds.append(-1)
-      else:
-        class_preds.append(1)
-    return class_preds
+    try:
+      preds = self.model.PredictLabels(X)
```
Comment (giovannt0) [inline estimator.py]: You can probably directly return `self.model.PredictLabels(X)` or `self.model.Predict(X)` rather than assigning the result to the variable `preds`, since it's unused afterwards

File: estimator.py [hunk 3]
```diff
@@ -109,16 +112,20 @@ def predict(self, X: np.array) -> np.array:
...
+    except ValueError:
+      preds = self.model.Predict(X)
+    return preds
+#    if not self.binary_classification:
```
Comment (giovannt0) [inline estimator.py]: Delete commented lines

File: model.py [hunk 4]
```diff
@@ -39,6 +40,7 @@ class GradientBoostingEnsemble:
   def __init__(self,
                nb_trees: int,
                nb_trees_per_ensemble: int,
+               n_classes: int = None,
```
Comment (giovannt0) [inline model.py]: `Optional[int] = None`

File: model.py [hunk 5]
```diff
@@ -125,29 +132,27 @@ def Train(self,
...
-          gradients = self.ComputeGradientForLossFunction(
-              y_ensemble, self.Predict(X_ensemble))  # type: ignore
+      # Compute sensitivity
+      delta_g = 3 * np.square(self.l2_threshold)
+      delta_v = min(self.l2_threshold / (1 + self.l2_lambda),
+                    2 * self.l2_threshold * math.pow(
+                      (1 - self.learning_rate), tree_index))
+
+      # Privacy budget allocated to each tree
+      tree_privacy_budget = np.divide(self.privacy_budget, nb_ensembles * self.loss_.K)
```
Comment (giovannt0) [inline model.py]: Why is `nb_ensembles` multiplied by `self.loss_.K` here?

File: model.py [hunk 6]
```diff
@@ -125,29 +132,27 @@ def Train(self,
...
         # Initialize the dataset and the gradients
         X_ensemble = np.copy(X)
         y_ensemble = np.copy(y)
-        prev_rmse = np.inf
+        prev_score = np.inf
         update_gradients = True
-        if tree_index > 0:
-          gradients = self.ComputeGradientForLossFunction(
-              y_ensemble, self.Predict(X_ensemble))
+        # gradient initialization will happen later in the per-class-loop
```
Comment (giovannt0) [inline model.py:162]: I see here you removed the gradients initialisation. Does it happen in the per-class-loop when using regression as well? If not, please add it back and adapt the `if` statement to work accordingly.

File: model.py [hunk 7]
```diff
@@ -185,57 +190,65 @@ def Train(self,
...
-          num_idx=self.num_idx)
-      tree.Fit(X_tree, gradients_tree)
-
-      # Add the tree to its corresponding ensemble
-      self.trees.append(tree)
-
-      rmse = np.sqrt(np.mean(np.square(y_test - self.Predict(X_test))))
-      if rmse >= prev_rmse:
+      k_trees = []
+      for kth_tree in range(self.loss_.K):
```
Comment (giovannt0) [inline model.py:205]: What is this loop for? Is it iterating over all trees? Something else? How different is it compared to the loop [here](https://github.com/giovannt0/dpgbdt/pull/1/files#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R138)? What is the expected behaviour for the regression task vs. the classification one? Perhaps add a few comments to explain what's happening.

File: model.py [hunk 8]
```diff
@@ -185,57 +190,65 @@ def Train(self,
...
+            self.loss_,
+            max_depth=self.max_depth,
+            max_leaves=self.max_leaves,
+            min_samples_split=self.min_samples_split,
+            use_bfs=self.use_bfs,
+            use_3_trees=self.use_3_trees,
+            cat_idx=self.cat_idx,
+            num_idx=self.num_idx)
+        tree.Fit(X_tree,
+                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble,
```
Comment (giovannt0) [inline model.py]: Maybe `(y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble` can be its own statement rather than within the function call for more readability?

File: model.py [hunk 9]
```diff
@@ -246,74 +259,77 @@ def Train(self,
...
-          if len(copy) == 0 or not child or not child.X.any():  # type: ignore
-            continue
-          # Apply exponential mechanism to find sub 3-node tree
-          probabilities = []
-          max_gain = -np.inf
-          for candidate_index, candidate in enumerate(copy):
-            if not candidate.root_node.X.any():
+    for index, k_three_tree in enumerate(k_trees):
+      k_trees_ = []
+      for k, three_tree in enumerate(k_three_tree):
```
Comment (giovannt0) [inline model.py]: Why is there an additional loop here now? Is that only for the classification task? Maybe add a comment explaining what's happening

File: model.py [hunk 10]
```diff
@@ -345,10 +362,23 @@ def SplitNode(self,
     rhs = np.where(rhs_op(node.X[:, index], value))[0]
 
     # Compute the associated predictions
-    lhs_prediction = (-1 * np.sum(node.gradients[lhs]) / (len(
-        node.gradients[lhs]) + self.l2_lambda))  # type: float
-    rhs_prediction = (-1 * np.sum(node.gradients[rhs]) / (len(
-        node.gradients[rhs]) + self.l2_lambda))  # type: float
+    if self.loss_.is_multi_class and len(node.gradients[lhs]) > 0:
```
Comment (giovannt0) [inline model.py]: Here I think we should introduce a function to make it a little easier to read. Something like `ComputePredictions(node: DecisionNode, indices: List[int], is_multi_class: bool)`. Then we can just call:

```
rhs_predictions = self.ComputePredictions(node, rhs, self.loss_.is_multi_class)
lhs_predictions = self.ComputePredictions(node, lhs, self.loss_.is_multi_class)
```

Makes it a little more readable

File: model.py [hunk 11]
```diff
@@ -392,23 +424,41 @@ def Predict(self, X: np.array) -> np.array:
     Returns:
       np.array: The predictions.
     """
-    predictions = np.sum(tree.Predict(X) for tree in self.trees)
+    predictions = np.sum([[tree.Predict(X) for tree in k_trees] for k_trees in self.trees], axis=0).T
```
Comment (giovannt0) [inline model.py:429]: Is `k_trees` an ensemble of tree per label to predict here? In the regression task shouldn't `predictions` be a `np.ndarray` rather than a list of `np.ndarray`?

File: model.py [hunk 12]
```diff
@@ -392,23 +424,41 @@ def Predict(self, X: np.array) -> np.array:
...
       y_pred (np.array): The predictions.
+      k (int): the class.
 
     Returns:
       (np.array): The gradient of the loss function.
     """
-    return np.multiply(-1, np.subtract(y, y_pred))
+    if self.loss_.is_multi_class:
+      y = (y == k).astype(np.float64)
+    return -self.loss_.negative_gradient(y, y_pred, k=k)
```
Comment (giovannt0) [inline model.py:466]: I'm not familiar with Scikit's `negative_gradient` method here; but if it's negative already as the name implies, do we need the extra `-` sign in front of `self.loss_`?

File: model.py [hunk 13]
```diff
@@ -830,17 +900,23 @@ def FindBestSplit(self,
...
       gradients (np.array): The gradients for the dataset instances.
+      y (np.ndarray): The dataset labels.
 
     Returns:
       float: The prediction for the leaf node
     """
-    prediction = (-1 * np.sum(gradients) / (len(
-        gradients) + self.l2_lambda))  # type: float
+    if self.loss.is_multi_class and len(gradients) > 0:
+      prediction = -1 * np.sum(gradients) * (self.loss.K - 1) / self.loss.K
```
Comment (giovannt0) [inline model.py]: Can you link here a paper / documentation that explains the reasoning for the predictions formula? Just for reference.

**Review by giovannt0 — COMMENTED** [review]
A few comments / questions. Otherwise it's looking good!

File: model.py (hunk 6 above)
Comment (kirschte) [reply model.py:162]: Regression output remains the same as before. :+1:

File: model.py (hunk 10 above)
Comment (kirschte) [reply model.py]: Great idea! There is actually another occurrence of this snipped; all of which I've now integrated into one function.

File: model.py (hunk 11 above)
Comment (kirschte) [reply model.py:429]: Per se your're right. However as `Predict` is also used internally, it is okay to abstract the regression onto a one-class regression (thus a list containing only one array).
In the user-friendly interface `DPGBDT` in `estimator.py` though, I now converted it back to its original shape, an `np.ndarray`.

File: model.py (hunk 6 above)
Comment (giovannt0) [reply model.py:162]: This `if` statement only is reached when there are multiple ensembles (not the case in `example.py`). Can you check if with e.g. 100 trees and 20 trees per ensemble there's no impact?
Also, irrespective of the output, why shouldn't we reinitialise the gradients here? Since these trees are independent from the previous ones and work on different data I'd expect the need of initialising the gradients again (note that the choice of initialisation is another topic, I'm not sure which is best in this case as this is not mentioned in the original paper)

File: estimator.py [hunk 14]
```diff
@@ -112,20 +112,18 @@ def predict(self, X: np.array) -> np.array:
       np.array: The predictions.
     """
     assert self.model
+    # try (multi-class) classification output first,
+    # ow fallback to the raw regression values
```
Comment (giovannt0) [inline estimator.py]: Small typo here: `ow` -> `or`

File: estimator.py [hunk 15]
```diff
@@ -112,20 +112,18 @@ def predict(self, X: np.array) -> np.array:
...
-#      else:
-#        class_preds.append(1)
-#    return class_preds
+      reg_preds = self.model.Predict(X).squeeze()  # shape: (n_samples,)
+      # binary classification is here conducted by regression
+      # and not by the deviance loss (could be improved later)
+      if not self.binary_classification:
+        return reg_preds
+      else:
+        return np.where(reg_preds < 0, -1, 1)
```
Comment (giovannt0) [inline estimator.py:126]: Sweet, thanks!

File: model.py [hunk 16]
```diff
@@ -231,8 +234,11 @@ def Train(self,
             use_3_trees=self.use_3_trees,
             cat_idx=self.cat_idx,
             num_idx=self.num_idx)
+        # in multi-class classification, the target has to be binary
+        # as each tree is a per-class regressor
         tree.Fit(X_tree,
-                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble,
+                 (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class
```
Comment (giovannt0) [inline model.py]: Maybe (y_ensemble == kth_tree).astype(np.float64) if self.loss_.is_multi_class else y_ensemble can be its own statement a few lines before rather than within the function call for more readability?

File: model.py (hunk 11 above)
Comment (giovannt0) [reply model.py:429]: Sounds good, thanks for the clarification!

**Review by giovannt0 — COMMENTED** [review]
A couple questions left, otherwise LGTM; thanks!

File: estimator.py (hunk 14 above)
Comment (kirschte) [reply estimator.py]: actually, ow means otherwise^^

File: model.py (hunk 6 above)
Comment (kirschte) [reply model.py:162]: Regression output remains the same even with multiple ensembles. :)
In fact, the gradients are reinitialized, just later on in [L201-203](https://github.com/giovannt0/dpgbdt/pull/1/files/1f66c420a24a00e3cbe7840aa3a680778c3047e6#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R201-R203). Sorry for the confusion, I just had to reorganize some code as the gradient initialization has to be done within the per-class loop and some constant parts shouldn't be.
Anyhow, it is a nice functionality to run multiple ensembles and imho it should still work ;)
